flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java

```diff
@@ -0,0 +1,458 @@
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+
+            TestingStopWithSavepoint sws =
+                    createStopWithSavepoint(ctx, new StateTrackingMockExecutionGraph());
+            ctx.setHowToHandleFailure(
+                    (ignore) -> Executing.FailureResult.canRestart(Duration.ZERO));
+
+            ctx.setExpectRestarting(assertNonNull());
+
+            sws.updateTaskExecutionState(createFailingStateTransition());
+        }
+    }
+
+    @Test
+    public void testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion()
+            throws Exception {
+        MockStopWithSavepointContext ctx = new MockStopWithSavepointContext();
+        TestingStopWithSavepoint sws = createStopWithSavepoint(ctx);
+
+        sws.onLeave(Canceling.class);
+
+        ctx.close();
+        assertThat(sws.getOperationCompletionFuture().isCompletedExceptionally(), is(true));
+    }
+
+    @Test
+    public void testExceptionalFutureCompletionAndStateTransitionOnLeaveAfterSavepointCompletion()
+            throws Exception {
+        MockStopWithSavepointContext ctx = new MockStopWithSavepointContext();
+        TestingStopWithSavepoint sws = createStopWithSavepoint(ctx);
+        ctx.setGlobalFailureHandler(sws);
+        ctx.setHowToHandleFailure((ignore) -> Executing.FailureResult.canRestart(Duration.ZERO));
+        ctx.setExpectRestarting(assertNonNull());
+        sws.getSavepointFuture().complete(createCompletedSavepoint());
+
+        sws.onLeave(Canceling.class);
+
+        ctx.close();
+        assertThat(sws.getOperationCompletionFuture().isCompletedExceptionally(), is(true));
+    }
+
+    @Test
+    public void testExceptionalSavepointCompletion() throws Exception {
+        MockStopWithSavepointContext ctx = new MockStopWithSavepointContext();
+        TestingStopWithSavepoint sws = createStopWithSavepoint(ctx);
+
+        sws.getSavepointFuture().completeExceptionally(new RuntimeException("Test error"));
+
+        ctx.close();
+        assertThat(sws.getOperationCompletionFuture().isCompletedExceptionally(), is(true));
+        assertThat(ctx.hadStateTransition, is(false));
+    }
+
+    @Test
+    public void testRestartOnTaskFailureAfterSavepointCompletion() throws Exception {
+        MockStopWithSavepointContext ctx = new MockStopWithSavepointContext();
+        TerminateExecutionFutureMockedExecutionGraph executionGraph =
+                new TerminateExecutionFutureMockedExecutionGraph();
+
+        TestingStopWithSavepoint sws = createStopWithSavepoint(ctx, executionGraph);
+        ctx.setGlobalFailureHandler(sws);
+
+        ctx.setHowToHandleFailure((ignore) -> Executing.FailureResult.canRestart(Duration.ZERO));
+
+        ctx.setExpectRestarting(assertNonNull());
+
+        sws.getSavepointFuture().complete(createCompletedSavepoint());
+        executionGraph.getExecutionFuture().complete(ExecutionState.FAILED);
```

Comment:
you are absolutely right. In general, relying on the futures for detecting task failure breaks consistency because it circumvents the normal control flow of adaptive scheduler.